doc/man/pvfs2-showcoll.1
.TH PVFS2-showcoll 1 2017-08-14
.SH NAME
\fBpvfs2-showcoll\fR \(en show server storage collection
.SH SYNOPSIS
\fBpvfs2-mkspace\fR [\fB\-s data_storage_space\fR]
[\fB\-m metadata storage space\fR] [\fB\-c collection_name\fR]
[\fB\-d dspace_handle\fR] [\fB\-vkh\fR] 
.SH DESCRIPTION
The
.B pvfs2-showcoll
utility displays the configuration of a particular storage collection.
This utility operates directly on the storage database and does not
utilize the server.
.PP
The options are as follows:
.IP -s
Specify the path to the data storage space.
.IP -m
Specify the path to the metadata storage space.
.IP -c
Specify the name of the collection.  Without this option, a list of
collections available will be displayed.
.IP -d
Specify the dspace handle to display.
.IP -v
Enable verbose output.
.IP -k
Display data in keyval spaces.
.IP -h
Display synopsis.
.SH ENVIRONMENT
.IP PVFS2_DEBUGFILE
If set to the path of a local file, redirect debug output to it.
.IP PVFS2_DEBUGMASK
Set the OrangeFS debug mask.  Possible masks are documented in
.BR pvfs2-set-debugmask ( 1 ) \& .
.SH BUGS
Please submit bug reports to devel@lists.orangefs.org
.SH SEE ALSO
.BR pvfs2tab ( 5 )


ChangeLog
-----------------------

OrangeFS Release ChangeLog
-----------------------

orangefs-2.10.0 (Git Branch v.2.10.0)
https://github.com/waltligon/orangefs.git
up to commit-id 7c28362b

New functionality:

* - general updates
	1. numerous autotools related m4 files are updated.
	2. code refactored in numerous places to eliminate
	   compiler warnings from new versions of gcc.
	3. documentation updates.

* - Server
	1. updated formatting to conform to standard in simple-stripe code.

* - Clients
	1. build liborangefs by default. liborangefs provides an orangefs
	   version of posix system calls.

      2. Windows Client
         * Retargeted Client to the user mode file system library "dokany"
           v1.5.1.1000. (See https://github.com/dokan-dev/dokany)
         * Retargeted build to Visual Studio 2022.
         * Updated Event Logging for modern versions of Windows.
         * Removed obsolete security modes.
         * Updated the executable installer to install dokany and necessary
           Visual C++ runtime libraries.
         * Updated OpenSSL.
         * Made several quality-of-life improvements.

* - bmi
	1. Eliminated IBV_EVENT_WQ_FATAL warnings
	2. Added a new RDMA module:
		* Utilizes the RDMA Communication Manager library (librdmacm),
		  abstracting transport-specific details out of connection management
		* Adds RoCE v2 support to BMI
	3. Added a new experimental InfiniBand module for future development:
		* Still a work in progress and minimally tested
		* Current experimental changes include:
			- Improved connection-based reference counting




* - general fixes
	1. Fixed warning about using fast networks and TCP in pvfs2-genconfig
	

* - Client fixes
        1. Windows Client:
	     * Fixed a failure to delete files in some cases.
	     * Fixed compatibility issues between dokany 1.5.1 and the prior version.
	     * Fixed issues with error codes and directory-handling.




-----------------------
Development 2.9.9
orangefs-2.9.8 (Git Branch v.2.9.x)
https://github.com/waltligon/orangefs.git
up to rev 65ab0d2

This version of OrangeFS includes many bug fixes.

	1. Disable usrint by default.
	2. Fix potential memory overwrites in various utilities.
		- Client-core
		- Showcoll
		- Mkspace
		- Xattr
		- Viewdist
		- Str-utils
	3. Eliminate compiler warnings.
		- State machines
		- Trove
	4. Improve parsing of config file.
	5. Update documentation flow.
	6. Updated readmes.
	7. Migrate documentation to hugo format.
	8. Update perf mon example.



New functionality:
	1.  Added RoCE v1 support to our IB suite.


* - Server 
	1.  Fixed problems with crdirent and rename when using
	    distributed directories.
	2.  Fixed intermittent problems with updating timestamps.

* - Clients
	1.  pvfs2-stat.c
		When the object is a symlink, we now display the string length
		of the target in the size field, making the output consistent
		with the linux kernel.  Corrected call to sys-lookup so that
		symbolic links are followed correctly when -L is specified on
		the command line.
	2.  System Interface
		1.  Fixed problems with crdirent and rename when using distributed
		    directories.
		2.  Fixed intermittent problems with updating timestamps.

* - BMI
	1.  Part of adding RoCE v1 support was to make the IB port used for
	    communication configurable.

* - Software Infrastructure

* - Testing
	1.  Generic tests, #467 and #477, of the xfstest suite were removed from our
	    testing because we do not support the functionality.

	2.  Updated our testing to use version 20170929 of LTP.  Updated
	    the list of testcases to be used.

* - Kernel Module




orangefs-2.9.7 (Subversion Branch v.2.9.x)
http://dev.orangefs.org/svn/orangefs/branches/v.2.9.x
up to rev 13115 in trunk


New functionality:
	1.  Added support for Oracle Linux 7.3 with one caveat:  Cannot use
	    OrangeFS as the backing filesystem for a user-defined loop device.
	    Support will be forth coming.

	2.  Added support for large LMDB databases using DBMaxSize config
	    parameter.  

	3.  Added scripts in the examples/profile.d directory that can be placed in 
	    /etc/profile.d.  These scripts add OrangeFS locations to
	    LIBRARY_PATH, C_INCLUDE_PATH, AND LD_LIBRARY_PATH.

	4.  Added support for Fedora 26.  Certificate security is not yet
	    compatible with OpenSSL 1.1 but work is in progress.

	5.  Official OrangeFS rpm is in the Fedora distribution.  The rpm will
	    be in either Fedora 27 or Fedora 28.  The rpm will also be in
	    their EPEL repository, allowing other RHEL distributions to access
	    it.

	6.  Added --with-lmdb to the configure script, so you can specify the
	    location of your LMDB installation.  Otherwise, LMDB delivered
	    with OrangeFS will be used.  NOTE: Either way, to use LMDB instead
	    of Berkely DB, you must specify --with-db-backend=lmdb on the
	    configure line.

	7.  pvfs2-genconfig will ask you where you want to put the server config 
	    file, with the default being the <installation directory>/etc/orangefs.conf.
	    You can still put the server config filename directly on the command line.

	8.  ACL processing is now done as specified by POSIX.


* - Server  
	* - Enhanced permission checking and job scheduling for tree
	    operations.
	* - Removed memory leaks from the getattr operation and restructured
	    keyval buffer management.
	* - Corrected problem with mgmt-create-root-dir related to having a
	    correct capability when calling tree-setattr.
	* - Fixed capability errors that may occur during mkdir.

src/apps/admin/pvfs2-showcoll.c
/*
 * (C) 2002 Clemson University and The University of Chicago
 *
 * See COPYING in top-level directory.
 */

#include "pvfs2-config.h"

#include <unistd.h>
#include <string.h>
#include <stdio.h>
#include <getopt.h>
#include <limits.h>
#include <string.h>
#include <ctype.h>
#include <stdlib.h>

#include "pvfs2.h"
#include "trove.h"
#include "pvfs2-attr.h"
#include "pvfs2-internal.h"


static char data_path[PATH_MAX] = "/tmp/pvfs2-test-space";
static char meta_path[PATH_MAX] = "/tmp/pvfs2-test-space";
static char config_path[PATH_MAX] = "/tmp/pvfs2-test-space";
static char collection[PATH_MAX];
static int verbose = 0, got_collection = 0, print_keyvals = 0, got_dspace_handle = 0;
TROVE_handle dspace_handle;

static int parse_args(int argc,
		      char **argv);
static int print_collections(void);
static int print_dspaces(TROVE_coll_id coll_id,
			 TROVE_handle root_handle,
                         TROVE_context_id trove_context,
			 int no_root_handle);
static int print_dspace(TROVE_coll_id coll_id,
			TROVE_handle handle,
                        TROVE_context_id trove_context);
static int print_dspace_keyvals(TROVE_coll_id coll_id,
				TROVE_handle handle,
                                TROVE_context_id trove_context,
				TROVE_ds_type type);
static char *type_to_string(TROVE_ds_type type);
static int print_keyval_pair(TROVE_keyval_s *key,
			     TROVE_keyval_s *val,
			     TROVE_ds_type type,
			     int sz);
static void print_object_attributes(struct PVFS_object_attr *a_p);
static void print_datafile_handles(PVFS_handle *h_p,
				   int sz);

int main(int argc, char **argv)
{
    int ret, count, no_root_handle = 0;
    TROVE_op_id op_id;
    TROVE_coll_id coll_id;
    TROVE_handle root_handle;
    TROVE_ds_state state;
    TROVE_keyval_s key, val;
    TROVE_context_id trove_context = -1;

    ret = parse_args(argc, argv);
    if (ret < 0) {
	fprintf(stderr,
		"%s: error: argument parsing failed; aborting!\n",
		argv[0]);
	return -1;
    }

    /* initialize trove, verifying storage space exists */
    ret = trove_initialize(TROVE_METHOD_DBPF,
                           NULL,
                           data_path,
                           meta_path,
                           config_path,
                           0);
    if (ret < 0) 
    {
        printf("Error from trove_initialize is %d.\n",ret);
	fprintf(stderr,
		"%s: error: trove initialize failed; aborting!\n",
		argv[0]);
	return -1;
    }

    if (verbose) fprintf(stderr,
			 "%s: info: initialized with storage spaces '%s', '%s' and '%s'.\n",
			 argv[0],
			 data_path, meta_path, config_path);

    /* if no collection was specified, simply print out the collections and exit */
    if (!got_collection)
    {
	ret = print_collections();
	if (ret != 0)
        {
	    fprintf(stderr,
		    "%s: error: collection iterate failed; aborting!\n",
		    argv[0]);
	    trove_finalize(TROVE_METHOD_DBPF);
	    return -1;
	}
	trove_finalize(TROVE_METHOD_DBPF);
	return 0;
    }

    /* a collection was specified.
     * - look up the collection
     * - find the root handle (or maybe show all the collection attribs?)
     * - print out information on the dataspaces in the collection
     */

    ret = trove_collection_lookup(TROVE_METHOD_DBPF,
                                  collection,
				  &coll_id,
				  NULL,
				  &op_id);
    if (ret != 1)
    {
	fprintf(stderr,
		"%s: error: collection lookup failed for collection '%s'; aborting!.\n",
		argv[0],
		collection);
	trove_finalize(TROVE_METHOD_DBPF);
	return -1;
    }

    if (verbose)
    {
       fprintf(stderr,
               "%s: info: found collection '%s'.\n",
	       argv[0],
	       collection);
    }

    ret = trove_open_context(coll_id, &trove_context);
    if (ret < 0)
    {
        fprintf(stderr, "trove_open_context failed\n");
        return -1;
    }

    /* find root handle */
    key.buffer = ROOT_HANDLE_KEYSTR;
    key.buffer_sz = ROOT_HANDLE_KEYLEN;
    val.buffer = &root_handle;
    val.buffer_sz = sizeof(root_handle);
    ret = trove_collection_geteattr(coll_id,
				    &key,
				    &val,
				    0,
				    NULL,
                                    trove_context,
				    &op_id);

    while (ret == 0)
    {
	ret = trove_dspace_test(coll_id,
                                op_id,
                                trove_context,
                                &count,
                                NULL,
                                NULL,
                                &state,
                                TROVE_DEFAULT_TEST_TIMEOUT);
    }
    if (ret != 1)
    {
	if (verbose)
        {
            fprintf(stderr,
	            "%s: warning: collection geteattr (for root handle) failed; aborting!\n",
	            argv[0]);
        }
        no_root_handle = 1;
    }

    /* TODO: NEED ITERATE FOR EATTRS? */

    /* TODO: GET A COUNT OF DATASPACES? */

    /* print basic stats on collection */
    if (no_root_handle)
    {
	fprintf(stdout,
		"Storage space %s and %s, collection %s (coll_id = %d, "
                "*** no root_handle found ***):\n",
		data_path,
		meta_path,
		collection,
		coll_id);
    }
    else
    {
	fprintf(stdout,
		"Storage space %s and %s, collection %s (coll_id = %d, "
                "root_handle = %s):\n",
		data_path,
		meta_path,
		collection,
		coll_id,
		PVFS_OID_str(&root_handle));
    }

    if (got_dspace_handle)
    {
        ret = print_dspace(coll_id, dspace_handle, trove_context);
    }
    else
    {
        ret = print_dspaces(coll_id,
                            root_handle,
                            trove_context,
                            no_root_handle);
    }

    trove_close_context(coll_id, trove_context);
    trove_finalize(TROVE_METHOD_DBPF);

    return 0;
}

static int parse_args(int argc, char **argv)
{
    int c;

    while ((c = getopt(argc, argv, "s:m:c:d:kvh")) != EOF) {
	switch (c)
        {
	    case 's':
		strncpy(data_path, optarg, PATH_MAX - 1);
		break;
            case 'm':
		strncpy(meta_path, optarg, PATH_MAX - 1);
                break;
	    case 'c': /* collection */
		got_collection = 1;
		strncpy(collection, optarg, PATH_MAX - 1);
		break;
	    case 'k':
		print_keyvals = 1;
		break;
	    case 'd':
		/* TODO: USE BIGGER VALUE */
		got_dspace_handle = 1;
                PVFS_OID_str2bin(optarg, &dspace_handle);
                break;
	    case 'v':
		verbose = 1;
		break;
	    case '?':
	    default:
		fprintf(stderr, "%s: error: unrecognized option '%c'.\n", argv[0], c);
	    case 'h':
		fprintf(stderr,
			"usage: pvfs2-showcoll [-s data_storage_space] [-m metadata storage space] [-c collection_name] [-d dspace_handle] [-v] [-k] [-h]\n");
		fprintf(stderr, "\tdefault storage space is '/tmp/pvfs2-test-space'.\n");
		fprintf(stderr, "\t'-v' turns on verbose output.\n");
		fprintf(stderr, "\t'-k' prints data in keyval spaces.\n");
		fprintf(stderr, "\t'-d' prints data for a single dspace only, given a handle (in hex).\n");
		fprintf(stderr, "\t'-h' prints this message.\n");
		fprintf(stderr, "\n\tWithout a collection name, a list of collections will be printed.\n");
		if (c == 'h') exit(0);
		return -1;
	}
    }
    return 0;
}

static int print_dspaces(TROVE_coll_id coll_id,
			 TROVE_handle root_handle,
                         TROVE_context_id trove_context,
			 int no_root_handle)
{
    int ret, count;
    TROVE_ds_position pos;
    TROVE_handle harray[64];
    TROVE_op_id op_id;
    TROVE_ds_state state;

    pos = TROVE_ITERATE_START;
    count = 64;

    while (count > 0)
    {
	int opcount;

	ret = trove_dspace_iterate_handles(coll_id,
					   &pos,
					   harray,
					   &count,
					   0 /* flags */,
					   NULL /* vtag */,
					   NULL /* user ptr */,
                                           trove_context,
					   &op_id);
	while (ret == 0)
        {
            ret = trove_dspace_test(coll_id,
                                    op_id,
                                    trove_context,
                                    &opcount,
                                    NULL,
                                    NULL,
                                    &state,
                                    TROVE_DEFAULT_TEST_TIMEOUT);
        }
	if (ret != 1)
        {
            return -1;
        }

	if (count > 0)
        {
	    int i;
	    for (i = 0; i < count; i++)
            {
		ret = print_dspace(coll_id, harray[i], trove_context);
	    }
	}
    }

    return 0;
}

static int print_dspace(TROVE_coll_id coll_id,
			TROVE_handle handle,
                        TROVE_context_id trove_context)
{
    int ret, opcount;
    TROVE_ds_attributes_s ds_attr;
    TROVE_op_id op_id;
    TROVE_ds_state state;

    ret = trove_dspace_getattr(coll_id,
			       handle,
			       &ds_attr,
			       0 /* flags */,
			       NULL /* user ptr */,
                               trove_context,
			       &op_id,
                               NULL);
    while (ret == 0)
    {
	ret = trove_dspace_test(coll_id,
                                op_id,
                                trove_context,
                                &opcount,
                                NULL,
                                NULL,
                                &state,
                                TROVE_DEFAULT_TEST_TIMEOUT);
    }
    if (ret != 1)
    {
        return -1;
    }
		
    fprintf(stdout,
	    "\t%s (dspace_getattr output: type = %s, b_size = %lld)\n",
	    PVFS_OID_str(&handle),
	    type_to_string(ds_attr.type),
	    (ds_attr.type == PVFS_TYPE_DATAFILE) ?
                    lld(ds_attr.u.datafile.b_size) : 0);

    if (print_keyvals)
    {
	ret = print_dspace_keyvals(coll_id,
                                   handle,
                                   trove_context,
                                   ds_attr.type);
	if (ret != 0)
        {
            return -1;
        }
    }

    return 0;
}

static char *type_to_string(TROVE_ds_type type)
{
    static char mf[] = "metafile";
    static char df[] = "datafile";
    static char sl[] = "symlink";
    static char di[] = "directory";
    static char dd[] = "dirdata";
    static char in[] = "internal";
    static char un[] = "unknown";

    switch (type)
    {
	case PVFS_TYPE_METAFILE:
	    return mf;
	case PVFS_TYPE_DATAFILE:
	    return df;
	case PVFS_TYPE_DIRDATA:
	    return dd;
	case PVFS_TYPE_SYMLINK:
	    return sl;
	case PVFS_TYPE_DIRECTORY:
	    return di;
        case PVFS_TYPE_INTERNAL:
            return in;
	default:
	    return un;
    }
}

static int print_dspace_keyvals(TROVE_coll_id coll_id,
				TROVE_handle handle,
                                TROVE_context_id trove_context,
				TROVE_ds_type type)
{
    int ret, count;
    TROVE_ds_position pos;
    TROVE_keyval_s key, val;
    TROVE_op_id op_id;
    TROVE_ds_state state;

    key.buffer    = malloc(256);
    key.buffer_sz = 256;
    key.read_sz   = 0;
    val.buffer    = malloc(65536);
    val.buffer_sz = 65536;
    val.read_sz   = 0;

    if (key.buffer)
    {
        memset(key.buffer,0,256);
    }
    if (val.buffer)
    {
        memset(val.buffer,0,65536);
    }

    if ( !(key.buffer && val.buffer) )
    {
        if (key.buffer)
        {
            free(key.buffer);
        }
        if (val.buffer)
        {
            free(val.buffer);
        }
        printf("%s: Unable to allocate memory.\n", __func__);
        return -1;
    }


    pos = TROVE_ITERATE_START;
    count = 1;

    while (count > 0)
    {
	int opcount;
        printf("%s:calling trove_keyval_iterate for %s.\n"
              ,__func__
              ,PVFS_OID_str(&handle));

	ret = trove_keyval_iterate(coll_id,
				   handle,
				   &pos,
				   &key,
				   &val,
				   &count,
				   0 /* flags */,
				   NULL /* vtag */,
				   NULL /* user ptr */,
                                   trove_context,
				   &op_id,
                                   NULL);

	while (ret == 0)
        {
            ret = trove_dspace_test(coll_id,
                                    op_id,
                                    trove_context,
                                    &opcount,
                                    NULL,
                                    NULL,
                                    &state,
                                    TROVE_DEFAULT_TEST_TIMEOUT);
        }
	if (ret != 1)
        {
            return -1;
        }

        printf("%s: count=%d\n", __func__, count);

	if (count > 0)
        {
            print_keyval_pair(&key, &val, type, 65536);
        }

        /* re-initialize key val */
        memset(key.buffer,0,256);
        memset(val.buffer,0,65536);
        key.buffer_sz = 256;
        val.buffer_sz = 65536;
        key.read_sz = 0;
        val.read_sz = 0;
    }

    free(key.buffer);
    free(val.buffer);

    return 0;
}

static void print_object_attributes(struct PVFS_object_attr *a_p)
{
    fprintf(stdout,
	    "(owner = %d, group = %d, perms = %o, objtype = %s)\n",
	    a_p->owner,
	    a_p->group,
	    a_p->perms,
	    type_to_string(a_p->objtype));
}

static void print_datafile_handles(PVFS_handle *h_p, int count)
{
    int i;

    for (i = 0; i < count && i < 10; i++)
    {
        fprintf(stdout, "\n\t\t\t\t%s", PVFS_OID_str(&h_p[i]));
    }

    if (i == 10)
    {
        fprintf(stdout, "...\n");
    }
    else
    {
        fprintf(stdout, "\n");
    }
}

static int print_keyval_pair(TROVE_keyval_s *key_p,
			     TROVE_keyval_s *val_p,
			     TROVE_ds_type type,
			     int sz)
{
    int key_printable = 0, val_printable = 0;

    if (isprint(((char *)key_p->buffer)[0]))
    {
        key_printable = 1;
    }
    if (isprint(((char *)val_p->buffer)[0]))
    {
        val_printable = 1;
    }

    if (key_printable && key_p->buffer_sz >= 64)
    {
        memset(&((char *)key_p->buffer)[64],0,1);
    }

    if (val_printable && val_p->buffer_sz >= 64)
    {
        memset(&((char *)key_p->buffer)[64],0,1);
    }

    if (!strncmp(key_p->buffer, "metadata", 9) &&
        val_p->read_sz == sizeof(struct PVFS_object_attr))
    {
	fprintf(stdout,
		"\t\t'%s' (%d): '%s' (%d) as PVFS_object_attr = ",
		(char *)key_p->buffer,
		key_p->read_sz,
		(char *)val_p->buffer,
		val_p->read_sz);
	print_object_attributes((struct PVFS_object_attr *) val_p->buffer);
    }
    else if (!strncmp(key_p->buffer, "dh", 17) &&
             val_p->read_sz % sizeof(PVFS_handle) == 0)
    {
	fprintf(stdout,
		"\t\t'%s' (%d): '%s' (%d) as handles = ",
		(char *)key_p->buffer,
		key_p->read_sz,
		val_printable ? (char *)val_p->buffer : "",
		val_p->read_sz);
	print_datafile_handles((PVFS_handle *)val_p->buffer,
                              val_p->read_sz / sizeof(PVFS_handle));
    }
    else if (type == PVFS_TYPE_DIRECTORY && !strncmp(key_p->buffer, "de", 3))
    {
	fprintf(stdout,
		"\t\t'%s' (%d): '%s' (%d) as a handle = %s\n",
		(char *) key_p->buffer,
		key_p->read_sz,
		val_printable ? (char *)val_p->buffer : "",
		val_p->read_sz,
                PVFS_OID_str((TROVE_handle *) val_p->buffer));
    }
    else if (type == PVFS_TYPE_DIRDATA && val_p->read_sz == 8)
    {
	fprintf(stdout,
		"\t\t'%s' (%d): '%s' (%d) as a handle = %s\n",
		(char *) key_p->buffer,
		key_p->read_sz,
		(char *) val_p->buffer,
		val_p->read_sz,
		PVFS_OID_str((TROVE_handle *)val_p->buffer));
    }
    else if (key_printable && !strncmp((char *)key_p->buffer,
                                       "user.pvfs2.meta_hint",20))
    {
        fprintf(stdout,
                "\t\t'%s' (%d): 0x%08llX (%d)\n"
                ,(char *)key_p->buffer
                ,(int)strlen((char*)key_p->buffer)
                ,*(unsigned long long *)val_p->buffer
                ,(int)sizeof(unsigned long));
    }
    else if (key_printable && !strncmp((char *)key_p->buffer,
                                       "user.pvfs2.mirror.mode",22))
    {
        fprintf(stdout,
                "\t\t'%s' (%d): %d (%d)\n"
                ,(char *)key_p->buffer
                ,(int)strlen((char*)key_p->buffer)
                ,*(unsigned int *)val_p->buffer
                ,(int)sizeof(unsigned int));
    }
    else if (key_printable && !strncmp((char *)key_p->buffer,
                                       "user.pvfs2.mirror.copies",24))
    {
        fprintf(stdout,
                "\t\t'%s' (%d): %d (%d)\n"
                ,(char *)key_p->buffer
                ,(int)strlen((char*)key_p->buffer)
                ,*(unsigned int *)val_p->buffer
                ,(int)sizeof(unsigned int));
    }
    else if (key_printable && !strncmp((char *)key_p->buffer,
                                       "user.pvfs2.mirror.handles",25))
    {
        fprintf(stdout,
                "\t\t'%s' (%d): '' (%d) as handles:"
                ,(char *)key_p->buffer
                ,(int)strlen((char*)key_p->buffer)
                ,(int)val_p->read_sz);
	print_datafile_handles((PVFS_handle *) val_p->buffer,
                               val_p->read_sz / sizeof(PVFS_handle));
    }
    else if (key_printable && val_printable)
    {
	fprintf(stdout,
		"\t\t'%s' (%d): '%s' (%d)\n",
		(char *) key_p->buffer,
		key_p->read_sz,
		(char *) val_p->buffer,
		val_p->read_sz);
    }
    else if (key_printable && !val_printable)
    {
	fprintf(stdout,
		"\t\t'%s' (%d): <data> (%d)\n",
		(char *) key_p->buffer,
		key_p->read_sz,
		val_p->read_sz);
    }
    else
    {
	fprintf(stdout,
		"\t\t<data> (%d): <data> (%d)\n",
		key_p->read_sz,
		val_p->read_sz);
    }
    return 0;
}

static int print_collections(void)
{
    int ret, count;
    TROVE_op_id op_id;
    TROVE_coll_id coll_id[32];
    TROVE_keyval_s name[32];
    int i;

    count = 32;
    for (i = 0; i < 32; i++)
    {
        char *coll_name;
        coll_name = malloc(PATH_MAX);
        if (coll_name == NULL)
            return -1;
        memset(coll_name,0,PATH_MAX);
        name[i].buffer    = coll_name;
        name[i].buffer_sz = PATH_MAX;
        name[i].read_sz   = 0;
    }

    fprintf(stdout, "Storage space %s and %s collections:\n",
	    data_path, meta_path);

    ret = trove_collection_iterate(TROVE_METHOD_DBPF, name, coll_id, &count,
        0 /* flags */, 0 /* vtag */, NULL /* user ptr */, &op_id);
    if (ret != 1) 
    {
        for (i = 0; i < 32; i++)
        {
            free(name[i].buffer);
        }
        return -1;
    }

    for (i = 0; i < count; i++)
    {
        fprintf(stdout, "\t%s (coll_id = %d)\n", (char *)name[i].buffer,
            coll_id[i]);
    }

    for (i = 0; i < 32; i++)
    {
        free(name[i].buffer);
    }

    fprintf(stdout, "\n");
    return 0;
}

/*
 * Local variables:
 *  c-indent-level: 4
 *  c-basic-offset: 4
 * End:
 *
 * vim: ts=8 sts=4 sw=4 expandtab
 */
